server/server_test.go

```diff
@@ -70,55 +71,82 @@ func TestServer_WithBaseContext(t *testing.T) {
 	}
 }
 
-func TestServer_Run(t *testing.T) {
-	// Create a new Server instance
-	server := NewServer(":0")
+func TestServer_WithReadinessChan(t *testing.T) {
+	// Create a new Server instance with a base context
+	ready := make(chan struct{})
+	server := NewServer(":0", WithReadinessChan(ready))
 
-	channel := mocks.NewMockChannel(t)
-	channel.EXPECT().Path().Return("/test")
-	channel.EXPECT().Handler().Return(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+	if server.ready == nil {
+		t.Error("Expected non-nil channel")
+	}
 
-	server.AddChannel(channel)
+	close(server.ready)
+	_, ok := <-ready
+	if ok {
+		t.Error("Expected closed channel")
+	}
+}
 
-	done := make(chan struct{})
+func TestServer_Run(t *testing.T) {
 
-	// Run the server
-	for i := 0; i < 2; i++ {
-		go func() {
-			err := server.Run()
-			switch err {
-			case nil:
-				close(done)
-			case ErrServerAlreadyRunning:
-				done <- struct{}{}
-			default:
-				t.Errorf("Got unexpected error: %v", err)
+	noOfReruns := []int{0, 1, 2, 5}
+
+	for _, run := range noOfReruns {
+		t.Run(fmt.Sprintf("%d times of calling Run", run), func(t *testing.T) {
+			// Create a new Server instance
+			ready := make(chan struct{})
+			server := NewServer(":0", WithReadinessChan(ready))
+
+			channel := mocks.NewMockChannel(t)
+			channel.EXPECT().Path().Return("/test")
+			channel.EXPECT().Handler().Return(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
+
+			server.AddChannel(channel)
+
+			// for signaling server stopped without error
+			done := make(chan struct{})
+
+			// Run the server
+			go func() {
+				err := server.Run()
+				switch err {
+				case nil:
+					close(done)
+				case ErrServerAlreadyRunning:
+					return
+				default:
+					t.Errorf("Got unexpected error: %v", err)
+				}
+			}()
+
+			// Wait for server to be ready
+			<-ready
```

Comment:
please cover this waiting with select that has timeout as alternative option, to not dead lock if ready channel is broken

```
select {
 			case <-ready:
 			case <-time.After(1 * time.Second):
 				t.Error("Expected server to start")
 			}
```